components/script/dom/webglrenderingcontext.rs

```diff
@@ -471,6 +545,99 @@ impl WebGLRenderingContext {
+
+        let mut flipped = Vec::<u8>::with_capacity(pixels.len());
+
+        for y in 0..height {
+            let flipped_y = height - 1 - y;
+            flipped.extend_from_slice(&pixels[(flipped_y * stride)..((flipped_y + 1) * stride)]);
+        }
+
+        flipped
+    }
+
+    fn premultiply_pixels(&self,
+                          format: TexFormat,
+                          data_type: TexDataType,
+                          pixels: Vec<u8>) -> Vec<u8> {
+        if !(self.texture_unpacking_settings.get().contains(PREMULTIPLY_ALPHA)) {
+            return pixels;
+        }
+
+        match (format, data_type) {
+            (TexFormat::RGBA, TexDataType::UnsignedByte) => {
+                let mut premul = Vec::<u8>::with_capacity(pixels.len());
+                for rgba in pixels.chunks(4) {
+                    premul.push(multiply_u8_pixel(rgba[0], rgba[3]));
+                    premul.push(multiply_u8_pixel(rgba[1], rgba[3]));
+                    premul.push(multiply_u8_pixel(rgba[2], rgba[3]));
+                    premul.push(rgba[3]);
+                }
+                premul
+            }
+            (TexFormat::LuminanceAlpha, TexDataType::UnsignedByte) => {
+                let mut premul = Vec::<u8>::with_capacity(pixels.len());
+                for la in pixels.chunks(2) {
+                    premul.push(multiply_u8_pixel(la[0], la[1]));
+                    premul.push(la[1]);
+                }
+                premul
+            }
+
+            // FINISHME: This is not valid for big endian -- we should
+            // be looking at the low bit of a u16, not the low bit of
+            // the first u8.
+            (TexFormat::RGBA, TexDataType::UnsignedShort5551) => {
+                let mut premul = Vec::<u8>::with_capacity(pixels.len());
+                for rgba in pixels.chunks(2) {
+                    if rgba[0] & 1 != 0 {
+                        premul.push(rgba[0]);
+                        premul.push(rgba[1]);
+                    } else {
+                        premul.push(0);
+                        premul.push(0);
+                    }
+                }
+                premul
+            }
+
+            (TexFormat::RGBA, TexDataType::UnsignedShort4444) => {
+                let mut premul = Vec::<u8>::with_capacity(pixels.len());
+                for rgba in pixels.chunks(2) {
+                    let a = ((rgba[0] & 0x0f) << 4) | rgba[0] & 0x0f;
+                    let b = ((rgba[0] & 0xf0) >> 4) | rgba[0] & 0xf0;
+                    let g = ((rgba[1] & 0x0f) << 4) | rgba[1] & 0x0f;
+                    let r = ((rgba[1] & 0xf0) >> 4) | rgba[1] & 0xf0;
+
+                    premul.push(rgba[0] & 0x0f |
+                                multiply_u8_pixel(b, a) & 0xf0);
+                    premul.push(multiply_u8_pixel(g, a) >> 4 |
+                                multiply_u8_pixel(r, a) & 0xf0);
+                }
+                premul
+            }
+
+            _ => pixels
```

Comment:
Is this also unreachable? If not, consider commenting which cases do we expect? Otherwise, mark as such?